sys/xtimer/xtimer_core.c

```diff
@@ -222,64 +210,61 @@ void xtimer_remove(xtimer_t *timer)
+    xtimer_t *timer = long_list_head;
 
-        while (curr_timer) {
-            uint32_t elapsed = *now - curr_timer->start_time;
-            if (curr_timer->offset <= elapsed) {
-                curr_timer->long_offset--;
-            }
-            curr_timer->offset -= elapsed;
-            curr_timer->start_time = *now;
+    while (timer) {
+        uint32_t elapsed = *now - timer->start_time;
 
-            if (!curr_timer->long_offset) {
-                assert(curr_timer == long_list_head);
+        if (timer->offset < elapsed) {
+            timer->long_offset--;
+        }
+        timer->offset -= elapsed;
+        timer->start_time = *now;
 
-                xtimer_t *trans_timer = curr_timer;
-                curr_timer = curr_timer->next;
-                _remove_timer_from_list(&long_list_head, trans_timer);
-                _add_timer_to_list(&timer_list_head, trans_timer);
-            }
-            else {
-                curr_timer = curr_timer->next;
-            }
+        if (!timer->long_offset) {
+            assert(timer == long_list_head);
+
+            _remove_timer_from_list(&long_list_head, timer);
+            _add_timer_to_list(&timer_list_head, timer);
+            timer = long_list_head;
+        }
+        else {
+            timer = timer->next;
         }
     }
 }
 
 /**
  * @brief update short timers' offsets and fire those that are close to expiry
  */
-static inline void _update_short_timers(uint32_t *now) {
-    while (timer_list_head) {
-        uint32_t elapsed = *now - timer_list_head->start_time;
-
-        if (timer_list_head->offset <= elapsed ||
-            timer_list_head->offset - elapsed < XTIMER_ISR_BACKOFF) {
-            /* make sure we don't fire too early */
-            while(_xtimer_now() - timer_list_head->start_time < timer_list_head->offset) {}
-
-            /* pick first timer in list */
-            xtimer_t *timer = timer_list_head;
+static inline void _update_short_timers(uint32_t *now)
+{
+    xtimer_t *timer = timer_list_head;
 
-            /* advance list */
-            timer_list_head = timer->next;
+    while (timer) {
+        assert(!timer->long_offset);
+        uint32_t elapsed = *now - timer->start_time;
+        if (timer->offset < elapsed || timer->offset - elapsed < XTIMER_ISR_BACKOFF) {
+            assert(timer == timer_list_head);
 
-            /* make sure timer is recognized as being already fired */
+            /* make sure we don't fire too early */
+            if (timer->offset > elapsed) {
+                while(_xtimer_now() - timer->start_time < timer->offset) {}
+            }
+            /* remove and fire timer */
             timer->offset = 0;
-            timer->long_offset = 0;
-            timer->next = NULL;
-
-            /* fire timer */
+            timer->start_time = 0;
+            _remove_timer_from_list(&timer_list_head, timer);
             _shoot(timer);
+            /* advance list */
+            timer = timer_list_head;
```

Comment:
Advanced list **after** shooting the timer since the timer callback function could change **timer_list_head**.

Reply:
does that work if the callback re-sets the timer? E.g.,

```
main() {
 xtimer_t t = { .callback = cb, .arg = &t };
 xtimer_set(&t, 100000);
  while(1);
}
void cb(void *arg) {
 // do something
 xtimer_t *t = arg;
 xtimer_set(t, 100000);
}
```

Reply:
Hmm.. Can you think of a reason? My logic is: Line 257 removes the timer from the list -> Line 258 adds the timer back to somewhere in the list and **timer_list_head** would be set accordingly -> Line 260 sets timer to the (maybe renewed) timer_list_head. So it seems fine.

Reply:
> So it seems fine.

Yes, now I get it.